Libplanet.Net/BlockCandidateTable.cs

```diff
@@ -18,79 +17,238 @@ namespace Libplanet.Net
+                            continue;
+                        }
+
+                        var newBlocks = branch.Blocks.ToList();
+                        Block<T> index = branch.Root;
+                        while (index.PreviousHash != null &&
+                               path.Blocks.Contains(index))
+                        {
+                            try
+                            {
+                                index = path.Blocks.Single(x => x.Hash.Equals(index.PreviousHash));
+                                newBlocks.Insert(0, index);
+                            }
+                            catch (Exception e) when (
+                                e is ArgumentException ||
+                                e is InvalidOperationException)
+                            {
+                                break;
+                            }
+                        }
+
+                        if (newBlocks.First().Hash.Equals(path.OldTip.Hash) ||
+                            newBlocks.First().Hash.Equals(path.BranchPoint.Hash))
+                        {
+                            var newBranch = new CandidateBranch<T>(newBlocks);
+
+                            longestBranch ??= newBranch;
+                            longestBranch = CompareBranch(newBranch, longestBranch);
+                        }
+                    }
+                }
+
+                if (longestBranch is { })
+                {
+                    newBag.Add(longestBranch);
+                    BestBranch = longestBranch;
+                }
+
+                Branches = newBag;
+            }
         }
 
-        public bool TryRemove(BlockHeader header)
+        /// <summary>
+        /// Removes every <see cref="CandidateBranch{T}"/> in <see cref="Branches"/>.
+        /// </summary>
+        public void Clear()
         {
-            return _blocks.TryRemove(header, out _);
+            Branches = new ConcurrentBag<CandidateBranch<T>>();
+            BestBranch = null;
         }
 
-        public void Cleanup(Func<IBlockExcerpt, bool> predicate)
+        public bool Any() => Branches.Any();
+
+        /// <summary>
+        /// Check every blocks are continuous by <see cref="Block{T}.PreviousHash"/>-wise each
+        /// other.
+        /// <remarks>
+        /// This method does not check whether branch is appendable to <see cref="BlockChain{T}"/>.
+        /// </remarks>
+        /// </summary>
+        /// <param name="branch">A series of <see cref="Block{T}"/>.
+        /// </param>
+        /// <returns>
+        /// returns true if <paramref name="branch"/> is continuous or its count is 1; returns false
+        /// if <paramref name="branch"/> is not continuous or its count is 0.
+        /// </returns>
+        private static bool AreBlocksContinuous(IEnumerable<Block<T>> branch)
         {
-            foreach (var blockHeader in _blocks.Keys)
+            var enumerable = branch as Block<T>[] ?? branch.ToArray();
```

Comment:
Why is casting needed here? 🤔

Reply:
I do casts this purely because of multiple enumeration warning. It seems unnecessary casting because CandidateBranch<T> isn't belong to any parallel situation. I'll try to fix with moving AreBlocksContinuous() into CandidateBranch<T> constructor.

Reply:
Solved with moving AreBlocksContinuous() to the constructor of `CandidateBranch<T>`